File: leetcode673_Number_of_Longest_Increasing_Subsequence.cpp
class Solution {
public:
    int findNumberOfLIS(vector<int>& nums) {
        const int SIZE = nums.size();
        long maxLen = 0, count = 0;
        vector<long> dp(SIZE + 1,1);
        vector<int> len(SIZE,1);
        dp[0] = 1;

        for(int i = 0;i < SIZE;i++)
            for(int j = i - 1;j >= 0;j--)
                if(nums[i] > nums[j]){
                    if(len[j] >= len[i]){
                        dp[i] = dp[j];
                        len[i] = len[j] + 1;
                    }
                    else if(len[j] + 1 == len[i] && dp[i] + dp[j] <= INT_MAX)
                        dp[i]+= dp[j];
                }

        for(int i = 0;i < SIZE;i++){
            if(maxLen < len[i]){
                maxLen = len[i];
                count = dp[i];
            }
            else if(maxLen == len[i] && count + dp[i] <= INT_MAX)
                count+= dp[i];
        }

        return count;
    }
};

// https://tinyurl.com/t6b35suy

File: leetcode673_Number_of_Longest_Increasing_Subsequence.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        final int SIZE = nums.length;
        long[] dp = new long[2001];
        int[] lens = new int[2001];
        long maxLen = 0, count = 0;
        
        for(int i = 0;i < 2001;i++){
            dp[i] = 1;
            lens[i] = 1;
        }
        
        for(int i = 0;i < SIZE;i++){
            for(int j = i - 1;j >= 0;j--){
                if(nums[j] < nums[i]){
                    if(lens[j] + 1 == lens[i] && dp[i] + dp[j] <= (long)Integer.MAX_VALUE){
                        dp[i]+= dp[j];
                    }
                    if(lens[j] + 1 > lens[i]){
                        lens[i] = lens[j] + 1;
                        dp[i] = dp[j];
                    }
                }
            }
            if(lens[i] == maxLen && count + dp[i] <= (long)Integer.MAX_VALUE)
                count+= (int)dp[i];
            if(lens[i] > maxLen){
                maxLen = lens[i];
                count = (int)dp[i];
            }
        }
        
        return (int)count;
    }
}

// https://tinyurl.com/t6b35suy